### PR TITLE
follow `returnTo` when hitting `/login` while already logged in

### PR DESCRIPTION
**Pull request by david-crespo**
I tried to go to RFD 704 in a browser where I was logged out (I use Chrome for Google Meet and nothing else). I landed on `https://rfd.shared.oxide.computer/login?returnTo=/rfd/0704`. I copied and pasted that URL into the browser I normally use because that's where I want to read it. I got taken to `/` instead of RFD 704. Turns out this is because we ignore `returnTo`. So in this change we stop doing that.

### CODE REVIEW

File: app/routes/login.tsx [hunk 1]
```diff
@@ -35,10 +35,11 @@ export const loader = async ({ request }: LoaderFunctionArgs) => {
...
   const emailResponse = url.searchParams.get('email')
 
-  // If we're already logged in, just return to the home page. This may occur
-  // while navigating back and forward or from their history
+  // If we're already logged in, honor returnTo if present (e.g. an auth-gated
+  // RFD redirected here), otherwise go home. This may also occur while
+  // navigating back/forward or from history.
   if (await getUserFromSession(request)) {
-    throw redirect('/')
+    throw redirect(returnTo ? sanitizeRedirect(returnTo) : '/')
```
Comment (david-crespo) [inline app/routes/login.tsx:42]: Same logic as the one existing `sanitizeRedirect` callsite.

https://github.com/oxidecomputer/rfd-site/blob/e05dd4605bff13b256e012664876dc1079e27621/app/services/auth.server.ts?plain=1#L165

File: app/services/auth.server.ts [hunk 2]
```diff
@@ -173,10 +173,15 @@ async function handleAuthenticationCallback(provider: string, request: Request)
...
 
-function sanitizeRedirect(path: string): string {
-  const decoded = decodeURIComponent(path)
+export function sanitizeRedirect(path: string): string {
+  let decoded
+  try {
+    decoded = decodeURIComponent(path)
+  } catch {
+    return '/'
+  }
```
Comment (david-crespo) [inline app/services/auth.server.ts:191]: This was just hardening against an edge case failure mode, classic GPT-5.5. Seems fine: low cost, covered in new unit test.

File: app/services/auth.server.ts [hunk 3]
```diff
@@ -173,10 +173,15 @@ async function handleAuthenticationCallback(provider: string, request: Request)
 
 export { auth, handleAuthenticationCallback }
 
-const RFD_PATH = /^\/rfd\/[0-9]{1,4}\??.*$/
+const RFD_PATH = /^\/rfd\/[0-9]{1,4}(?:\/discussion)?(?:\?.*)?$/
```
Comment (fakemonster) [inline app/services/auth.server.ts]: the previous regex was effectively `/^\/rfd\/[0-9]{1,4}/` (`\??.*` simplifies to `.*`, `.*$` in practice simplifies to nothing at all), so now this regex DOES explicitly match for RFD paths and not just "things that start with an RFD path". good fix

but: why does /discussion make the cut here, and not pdf, raw, etc? i could make an argument to tolerate any sub-path at all, honestly (as long as you see a slash after the rfd digits, it "looks right enough" within this context). would help to have some elaboration in the tests (or a commit message) of why discussion gets the special case, and perhaps to move the /raw example in the tests to the `'rejects non-RFD paths'` test, since it's not really malformed. if we really do want to reject all other rfd sub-paths, that is

Comment (david-crespo) [reply app/services/auth.server.ts]: I saw this go by as a robot choice but didn't really think about why. Turns out it had a decent reason. `/discussion` gets included is because it's the only subpath the app currently constructs as a `returnTo`. 

https://github.com/oxidecomputer/rfd-site/blob/e05dd4605bff13b256e012664876dc1079e27621/app/routes/rfd.%24slug.discussion.tsx#L25-L28

`/raw` defines a `returnTo`, but it goes to the RFD page, not `/raw`:

https://github.com/oxidecomputer/rfd-site/blob/e05dd4605bff13b256e012664876dc1079e27621/app/routes/rfd.%24slug.raw.tsx#L36-L38

and `/pdf` just 404s if you hit it while logged out on a private RFD.

https://github.com/oxidecomputer/rfd-site/blob/e05dd4605bff13b256e012664876dc1079e27621/app/routes/rfd.%24slug.pdf.tsx#L19-L22

This status quo doesn't really make sense — `/raw` and `/pdf` should also construct `returnTo` using their actual paths. So I fixed that in 8b7f1a5b67fccf72aa21be3bc0aae89882c9bec2 and loosened the regex to allow any subpath.

**Review by fakemonster — APPROVED** [review]

**Comment by david-crespo** [comment]
Confirmed it works in prod.
